packages/shared/src/util/index.ts

```diff
@@ -20,16 +20,34 @@ export * from "./saved-groups";
 
 export function getAffectedEnvsForExperiment({
   experiment,
+  linkedFeatures,
 }: {
   experiment: ExperimentInterface | ExperimentInterfaceStringDates;
+  linkedFeatures?: FeatureInterface[];
 }): string[] {
   // Visual changesets are not environment-scoped, so it affects all of them
-  if (experiment.hasVisualChangesets || experiment.hasURLRedirects)
+  // Also fallback to all envs if linkedFeatures is undefined, but the experiment does actually have linked features
+  if (
+    experiment.hasVisualChangesets ||
+    experiment.hasURLRedirects ||
+    (!linkedFeatures && !!experiment.linkedFeatures?.length)
+  )
     return ["__ALL__"];
 
-  // TODO: get actual environments for linked feature flags. We are being overly conservative here
-  if (experiment.linkedFeatures && experiment.linkedFeatures.length > 0) {
-    return ["__ALL__"];
+  if (linkedFeatures?.length) {
+    const envs: string[] = [];
+
+    linkedFeatures.forEach((linkedFeature) => {
+      const envSettings = linkedFeature.environmentSettings;
+      if (!linkedFeature.archived) {
+        for (const env in envSettings) {
```

Comment:
We only want to include environments where the experiment rule exists.  There's a helper method `getMatchingRules` we should use for this instead of trying to loop through envs manually.

Also, this method will need to know the list of environments from the organization. We can't rely on the keys in `environmentSettings` because some of those environments might have been deleted in the org.